src/eloelo/elodisco/dota_bot.rs

```diff
@@ -131,74 +144,80 @@ impl DotaBot {
             .join(",\n")
     }
 
-    /// Assigns random heroes in a way that minimizes chances for conflicts
-    fn assign_random_heroes<'a>(
-        &self,
-        users: &'a [DiscordUsername],
-    ) -> HashMap<&'a DiscordUsername, Vec<&Hero>> {
-        let mut hero_pools: Vec<(&DiscordUsername, Vec<&Hero>)> = self.make_hero_pools(users);
-        // Sort players by pool length
-        hero_pools.sort_by_key(|v| (v.1.len(), rand::random::<i32>())); // TODO: randomize order of same-size users
-
-        // Select one hero per player, banning the selected hero along the way
-        // (Ignore when a player's pool is empty)
-        let mut taken: HashSet<&Hero> = HashSet::new();
-        let mut selections: HashMap<&DiscordUsername, Vec<&Hero>> = HashMap::new();
-        for _ in 0..3 {
-            for (user, hero_pool) in &mut hero_pools {
-                hero_pool.shuffle(&mut rand::thread_rng());
-                while let Some(hero) = hero_pool.pop() {
-                    if taken.contains(hero) {
-                        continue;
-                    }
-                    taken.insert(hero);
-                    selections.entry(user).or_default().push(hero);
-                    break;
-                }
-            }
+    fn user_hero_pool(&self, username: &DiscordUsername) -> Vec<Hero> {
+        let user_state = self.state.get(username).expect("discord user state");
+        let pool: Vec<_> = if !user_state.allowed_heroes.is_empty() {
+            user_state.allowed_heroes.iter().cloned().collect()
+        } else {
+            self.heroes
+                .difference(&user_state.banned_heroes)
+                .cloned()
+                .collect()
+        };
+        let should_avoid_duplicates = match user_state.last_match_date {
+            Some(date) => Local::now() < date + chrono::Duration::days(1),
+            None => false,
+        };
+        let has_enough_heroes_to_avoid_duplicates = pool.len() >= 3;
+        if should_avoid_duplicates
+            && has_enough_heroes_to_avoid_duplicates
+            && !user_state.duplicate_heroes_opt_out
+        {
+            pool.into_iter()
+                .filter(|h| !user_state.last_match_heroes.contains(h.as_str()))
+                .collect()
+        } else {
+            pool
         }
-        selections
     }
 
-    fn make_hero_pools<'a>(
-        &self,
-        users: &'a [DiscordUsername],
-    ) -> Vec<(&'a DiscordUsername, Vec<&Hero>)> {
-        users
+    fn make_hero_pools(&self, match_start: &MatchStart) -> Vec<(PlayerInfo, Vec<Hero>)> {
+        debug!("dota bot state: {:?}", self.state);
+        let create_player_info = |id: &PlayerId, elo: i32, is_radiant: bool| -> PlayerInfo {
+            let discord_username = match_start
+                .player_db
+                .get(id)
+                .unwrap()
+                .discord_username()
+                .unwrap();
```

Comment:
Cant we fall back to PlayerId if discord username is not available? I don't think we will have anyone without discord account but it would be better if the app did not crash when e.g. testing.